Metadata import

jojoelfe opened the pull request:
# Description

The goal of this pull request is to import metadata from SerialEM mdoc files into the database. The main use-case for this is to use information like stage position, tilt-Angle, and beam-Image shift to automate assembling tilt-series from movies pre-processed in cisTEM. The metadata might also be useful for single-particle processing (refining beam-tilt as a function of beam-image shift).

The metadata for each movie is stored in a new table called MOVIE_ASSETS_METADATA, which has columns to store the movie asset id, a text field describing the kind of metadata (for now only SerialEM mdocs from movies), some key information, and the actual mdoc file, converted to json, as a string.
```
MOVIE_ASSET_METADATA_ID  MOVIE_ASSET_ID  METADATA_SOURCE       TILT_ANGLE  STAGE_POSITION_X  STAGE_POSITION_Y  STAGE_POSITION_Z  IMAGE_SHIFT_X  IMAGE_SHIFT_Y  EXPOSURE_DOSE  ACQUISITION_TIME  CONTENT_JSON                                                                                                                                                                                                                                                                                                                                 
-----------------------  --------------  --------------------  ----------  ----------------  ----------------  ----------------  -------------  -------------  -------------  ----------------  --------------------------------------------------------------------------------------------------------------------                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
1                        1               serialem_frames_mdoc  -32.9895    115.454           237.11            -14.6724          0.211688       -0.189643      3.38387        1396089257        {"ImageShift" : "0.211688 -0.189643","TiltAngle" : "-32.9895","Defocus" : "-6.52608","OperatingMode" : "1","UsingCDS" : "0","RotationAngle" : "178.414","StagePosition" : "115.454 237.11","Binning" : "0.5","ExposureDose" : "3.38387","UncroppedSize" : "-2880 -2046","GainReference" : "SuperRef_s_s001_20_00000_-15.0_Sep22_20.09.30.dm4"
2                        2               serialem_frames_mdoc  -2.99563    115.455           237.076           -14.6724          0.273956       -0.341381      3.38387        1396089268        {"ImageShift" : "0.273956 -0.341381","TiltAngle" : "-2.99563","Defocus" : "-7.29899","OperatingMode" : "1","UsingCDS" : "0","RotationAngle" : "178.414","StagePosition" : "115.455 237.076","Binning" : "0.5","ExposureDose" : "3.38387","UncroppedSize" : "-2880 -2046","GainReference" : "SuperRef_s_s001_20_00000_-15.0_Sep22_20.09.30.dm4
```



# I have rebased my feature branch to be current with the master branch using to minimize conflicts and headaches

- [x] yes
- [ ] no

# Which compilers were tested

- [ ] g++
- [x] icpc
- [ ] clang
- [ ] other (please specify)

# These changes are isolated to the

- [x] gui
- [x] core library
- [ ] gpu core library
- [ ] program it modifies

# How has the functionality been tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [x] Tested manually from GUI
- [ ] Tested manually from CLI
- [ ] Passed console tests
- [ ] Passed samples functional testing
- [ ] other (please specify)

# Checklist:

- [x] I have not changed anything that did not need to be changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, (***w.r.t. why***), particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/bHimes/cisTEM_docs) {Ok to pass for now}
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

bHimes commented on src/core/database.cpp:
```diff
@@ -819,6 +819,8 @@ bool Database::CreateAllTables()
+	success = CreateMovieMetadataTable();
+	CheckSuccess(success);
 
 	return success;
```
@timothygrant80 I'm not sure what the return value is used for, but this is only going to report on the final value for success, right? Should there be an 'allPassed' or something?

timothygrant80 replied:
CheckSuccess is a macro, it will immediately return false from the function if anything fails.  i guess it might be more obvious if the final line said return true, but the effect is the same.

bHimes reviewed (commented):
All in all, I think this seems to be good. Did you have to do anything special to get the PR template to activate, or did it pop up when you created the PR?

jojoelfe commented:
> 
> 
> All in all, I think this seems to be good. Did you have to do anything special to get the PR template to activate, or did it pop up when you created the PR?

I copy-pasted the .md file, it does not come up when creating the PR.

jojoelfe commented on src/gui/ProjectX_gui.h:
```diff
@@ -105,7 +105,7 @@ class MainFrame : public wxFrame
 		wxMenu* HelpMenu;
 
-		// Virtual event handlers, overide them in your derived class
+		// Virtual event handlers, override them in your derived class
```
Somewhere in there should be the addition of a new checkbox to enable the import of metadata. But yes, most changes are probably due to the mismatch of wxFormbuilder version in the autogenerated code.

jojoelfe commented on src/gui/ProjectX_gui.cpp:
```diff
@@ -320,7 +320,7 @@ AbInitio3DPanelParent::AbInitio3DPanelParent( wxWindow* parent, wxWindowID id, c
 	PleaseCreateRefinementPackageText = new wxStaticText( InputParamsPanel, wxID_ANY, wxT("Please create a refinement package (in the assets panel) in order to perform a 3D refinement."), wxDefaultPosition, wxDefaultSize, 0 );
 	PleaseCreateRefinementPackageText->Wrap( -1 );
-	PleaseCreateRefinementPackageText->SetFont( wxFont( wxNORMAL_FONT->GetPointSize(), wxFONTFAMILY_DEFAULT, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_BOLD, false, wxEmptyString ) );
+	PleaseCreateRefinementPackageText->SetFont( wxFont( wxNORMAL_FONT->GetPointSize(), wxFONTFAMILY_DEFAULT, wxFONTSTYLE_NORMAL, wxFONTWEIGHT_NORMAL, false, wxEmptyString ) );
```
No, I made no changes, must be due to wxFormbuilder version.

timothygrant80 replied:
weird, i presume that the font is no longer bold..

timothygrant80 replied:
@jojoelfe - When you go to something with bold text - for example the show job details in one of the result panels - is the text bold in your compiled cisTEM version?

jojoelfe replied:
No, it is not... 
I'll try to hunt down what happened. I just downloaded the Oct2018 version of wxformbuilder, so I can check if it is a version issue.

timothygrant80 replied:
Ok thanks.. I think we want to keep those things bold and going and changing them all will not be fun!

jojoelfe commented on src/gui/MovieImportDialog.cpp:
```diff
@@ -529,6 +529,11 @@ void MyMovieImportDialog::ImportClick( wxCommandEvent& event )
 
+		// list of added movies for subsequent metadata insertion
+		// 
+		wxArrayInt list_asset_id;
```
The advantage of wxArray is that it will compile everywhere, no matter if and what version of STL is available. However, it seems to me like nowadays that should not be an issue anymore. @timothygrant80 do you think STL would be ok?

bHimes replied:
It's only a minor preference, and since wxArray is essentially a wrapper to std::vector anyway, maybe it is better to leave it as is, especially in the gui code. I didn't think about it impacting compiling. Though I will point out, the GPU code requires standard 17

timothygrant80 replied:
it is true that by now std::vector should work on anything we care about.  At some point we could change all the arrays to vectors, but I think we should be consistent.

jojoelfe commented on src/core/database.h:
```diff
@@ -196,7 +200,7 @@ public :
 																																														"MASK_ASSET_ID", "MASK_EDGE_WIDTH", "OUTSIDE_MASK_WEIGHT", "SHOULD_LOWPASS_OUTSIDE_MASK", "MASK_FILTER_RESOLUTION");};
 
-
+	bool CreateMovieMetadataTable() { return CreateTable("MOVIE_ASSETS_METADATA", "Ptt", "MOVIE_ASSET_ID", "METADATA_TYPE", "JSON"); }
```
- metadatatype should store a string that indicates the type of metadata. For now the only type is SerialEM FrameSet mdocs, but in the future it could include EPU XML files or SerialEM tilt-series mdocs. 
- JSON holds the actual metadata in JSON format. An alternative approach would be to create a column for every field of the metadata, but that would be less flexible if new fields are included in the future and for accommodation metadata from different programs.

bHimes replied:
> 
> 
>     * metadatatype should store a string that indicates the type of metadata. For now the only type is SerialEM FrameSet mdocs, but in the future it could include EPU XML files or SerialEM tilt-series mdocs.

Okay, then maybe something like **METADATA_SOURCE_SOFTWARE** or **METADATA_FORMAT**

> 
>     * JSON holds the actual metadata in JSON format. An alternative approach would be to create a column for every field of the metadata, but that would be less flexible if new fields are included in the future and for accommodation metadata from different programs.

Oh cool! I like the flexibility that provides. How will you deal with metadata that is/isn't present? The database will just store and return whatever is there, so it won't break (which is rad) so will it just be program specific error handling?

bHimes replied:
Also with the name, perhaps **JSON_TABLE** or something to indicate what it is given that JSON itself is a data storage *format*

timothygrant80 replied:
So I think if you plan to support other file types in the future, you may want to have the important points as database columns.  Otherwise each program will have to check the format and read things in a different way depending on that etc.  Adding columns later isn't necessarily a problem.  Why don't you keep the whole JSON string, but also specifically fill columns for the data you think you will need in the near future?

bHimes commented:
> 
> 
> > All in all, I think this seems to be good. Did you have to do anything special to get the PR template to activate, or did it pop up when you created the PR?
> 
> I copy-pasted the .md file, it does not come up when creating the PR.

Got it, I'll look into this more, it was nice to see it. It wasn't too much of a drag to complete?

timothygrant80 commented on src/gui/ProjectX_gui.h:
```diff
@@ -1,5 +1,5 @@
 ///////////////////////////////////////////////////////////////////////////
-// C++ code generated with wxFormBuilder (version Oct 26 2018)
+// C++ code generated with wxFormBuilder (version 3.10.0-4761b0c5)
```
I think it should be safe to rm 35.  We can switch to 3.10 now, I think that should be fine.

bHimes commented:
Does the new commit mean you've switched wxFormBuilder versions?

jojoelfe commented:
> Does the new commit mean you've switched wxFormBuilder versions?

Yes, the diff is much cleaner using the older version and there is no issue with the fonts. The update should be separate from this pull request. I'll squash the commits later (or can one do that during merging?), so the history will be clean.

bHimes commented:
> 
> 
> > Does the new commit mean you've switched wxFormBuilder versions?
> 
> Yes, the diff is much cleaner using the older version and there is no issue with the fonts. The update should be separate from this pull request. I'll squash the commits later (or can one do that during merging?), so the history will be clean.

Squashing can be done on merge, but I don't think it allows you to edit commit messages etc, it just puts all commits into one. I think that is clean enough for this case. Since no one else has worked on your branch, it would be safe to locally do an interactive rebase (git rebase -i HEAD~4) for example would allow you to keep, squash or combine the last 4 commits. The commit messages etc can be kept (squash) edited (e) or discarded while keeping the contents of the commit.

Let's set a wxFormbuilder module up at UMMS that points to whatever version is used collectively, to ensure we all stay on the same page. Sound good?

timothygrant80 commented:
This looks good now - it is marked as a work in progress for me - are you still working on it, or should I approve it?

jojoelfe commented:
I am still working on it. Mostly want to rewrite the database format a bit following your recommendations.

jojoelfe commented:
@timothygrant80 I think this is ready now. Take a look, I introduced an asset class (which will be helpful for reading the data later) and also save some of the most important info in separate columns. I updated the description on top with an example of how the results look like.

bHimes commented on src/core/database.cpp:
```diff
@@ -1194,12 +1194,23 @@ void Database::EndMovieAssetInsert()
-void Database::AddNextMovieAssetMetadata(int movie_asset_id,  wxString metadata_type, wxString json)
-{
-	AddToBatchInsert("itt", movie_asset_id, metadata_type.ToUTF8().data(), json.ToUTF8().data());
+	BeginBatchInsert("MOVIE_ASSETS_METADATA", 11, "MOVIE_ASSET_ID", "METADATA_SOURCE", "CONTENT_JSON", "TILT_ANGLE", "STAGE_POSITION_X", "STAGE_POSITION_Y", "STAGE_POSITION_Z", "IMAGE_SHIFT_X", "IMAGE_SHIFT_Y", "EXPOSURE_DOSE", "ACQUISITION_TIME");
```
I'm not sure "EXPOSURE_DOSE" makes sense to me. Suggestions from existing fields include TOTAL_EXPOSURE, PRE_EXPOSURE, EXPOSURE_PER_FRAME, EXPOSURE_AMOUNT (though this one is also a bit wonky as the exposure we refer to is actual a number density (e/A^2) and an "amount" should just be a quantity.)

Less opposed to AQUISITION_TIME, however, this could be the time of the aquisition (I assume this is intended as it could be used to sort exposures) or it could read as the duration of each exposure (seconds the beam was on during aquire.) Suggest: TIME_OF_AQUISITION

bHimes replied:
...or TIME_OF_RECORDING or something

jojoelfe replied:
In other places in cisTEM does TOTAL_EXPOSURE include the PRE_EXPOSURE? Or is it the product of EXPOSURE_PER_FRAME and the number of frames?

bHimes replied:
> 
> 
> In other places in cisTEM does TOTAL_EXPOSURE include the PRE_EXPOSURE? Or is it the product of EXPOSURE_PER_FRAME and the number of frames?

TOTAL_EXPOSURE is used to asses radiation damage in simulate.cpp and reconstruct_3d.cpp, and is used to indicate the exposure the sample has received = pre-expsoure + exposure_per_frame*n_frames. In most cases, this is just pre_exposure_exposure_per_frame as n_frames will be 1 where calculated.

bHimes reviewed (commented):
Class looks good to me - only minor name suggestions

bHimes commented:
I'm approving this since all conversations seem to be well resolved. This way @twagner9 can make his modifications after updating to the changes in this PR.
